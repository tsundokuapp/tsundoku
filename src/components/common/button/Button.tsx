'use client';

import { cn } from '@/helpers/twUtils';

interface IButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode;
  onClick: () => void;
  className?: string;
  icon?: JSX.Element;
  sideIcon?: 'left' | 'right';
}

export const Button = ({
  children,
  onClick,
  className,
  icon,
  sideIcon = 'left',
  ...props
}: IButtonProps) => {
  return (
    <button
      onClick={onClick}
      className={cn(
        'flex w-full max-w-[180px] items-center justify-center rounded-lg border-2 bg-white px-4 py-2 font-semibold text-textLight transition-colors hover:bg-hoverBgLight disabled:cursor-not-allowed disabled:bg-slate-300 disabled:opacity-60 dark:border-slate-700 dark:bg-slate-900 dark:text-textDark dark:hover:bg-hoverBgDark disabled:dark:opacity-60 disabled:dark:hover:bg-slate-900',
        className,
      )}
      {...props}
    >
      {icon && sideIcon === 'left' && icon}
      {children}
      {icon && sideIcon === 'right' && icon}
    </button>
  );
};
