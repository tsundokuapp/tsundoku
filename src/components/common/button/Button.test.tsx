import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import { Button } from '@/components/common/button/Button';

describe('Testes para o componente <Button />', () => {
  describe('Testes de renderização', () => {
    it('Deve renderizar Button corretamente', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button).toBeInTheDocument();
    });

    it('Deve renderizar Button com texto correto', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button).toHaveTextContent('clique');
    });

    it('Deve renderizar Button com ícone na esquerda', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
          icon={<span>Icon</span>}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByText('clique');
      const Icon = screen.getByText('Icon');

      // Then
      expect(button.contains(screen.getByText('Icon'))).toBeTruthy();
      expect(Icon.nextSibling).toHaveTextContent('clique');
    });

    it('Deve renderizar Button com ícone na direita', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
          icon={<span>Icon</span>}
          sideIcon="right"
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByText('clique');
      const Icon = screen.getByText('Icon');

      // Then
      expect(button.contains(screen.getByText('Icon'))).toBeTruthy();
      expect(Icon.previousSibling).toHaveTextContent('clique');
    });

    it('Deve renderizar Button com classe customizada', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
          className="custom-class"
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button).toHaveClass('custom-class');
    });

    it('Deve aceitar e sobrepor a classe padrão de Button através da prop style', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
          style={{ backgroundColor: '#f1f' }}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button.style.backgroundColor).toBe('rgb(255, 17, 255)');
    });
  });

  describe('Testes de interação e comportamento', () => {
    it('Deve chamar a função do onClick quando clicado', () => {
      const handleClick = jest.fn();
      // Given
      render(
        <Button
          onClick={() => {
            handleClick();
          }}
        >
          clique
        </Button>,
      );

      // When
      fireEvent.click(screen.getByRole('button'));

      // Then
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('Deve tornar Button inacessível quando desabilitado', () => {
      const handleClick = jest.fn();
      // Given
      render(
        <Button
          onClick={() => {
            handleClick();
          }}
          disabled
        >
          clique
        </Button>,
      );

      // When
      fireEvent.click(screen.getByRole('button'));

      // Then
      expect(handleClick).not.toHaveBeenCalled();
    });
  });

  describe('Testes de acessibilidade', () => {
    it('Deve ter o atributo role igual a button', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button).toBeInTheDocument();
    });

    it('Deve ter o atributo tabindex igual a 0', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button).toHaveAttribute('tabindex', '0');
    });

    it('Deve aplicar o atributo arial-label quando passado', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
          aria-label="Botão de clique"
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');

      // Then
      expect(button).toHaveAttribute('aria-label', 'Botão de clique');
    });

    it('Dever ser possível focar no botão', () => {
      // Given
      render(
        <Button
          onClick={() => {
            /**/
          }}
        >
          clique
        </Button>,
      );

      // When
      const button = screen.getByRole('button');
      button.focus();

      // Then
      expect(button).toHaveFocus();
    });
  });

  describe('Teste de snapshot', () => {
    it('Deve manter o snapshot do Button', () => {
      // Given
      const { container } = render(
        <Button
          onClick={() => {
            /**/
          }}
        >
          clique
        </Button>,
      );

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
