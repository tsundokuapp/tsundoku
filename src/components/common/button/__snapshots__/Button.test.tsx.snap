// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testes para o componente <Button /> Teste de snapshot Deve manter o snapshot do Button 1`] = `
<div>
  <button
    class="flex w-full max-w-[180px] items-center justify-center rounded-lg border-2 bg-white px-4 py-2 font-semibold text-textLight transition-colors hover:bg-hoverBgLight focus:border-primary disabled:cursor-not-allowed disabled:bg-slate-300 disabled:opacity-60 dark:border-slate-700 dark:bg-slate-900 dark:text-textDark dark:hover:bg-hoverBgDark focus:dark:border-primary disabled:dark:opacity-60 disabled:dark:hover:bg-slate-900"
    tabindex="0"
  >
    clique
  </button>
</div>
`;
